lib/workflow/stores/mongo.js

```diff
@@ -677,11 +695,22 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
 
         var query = {
             graphId: data.graphId,
-            reachable: true
+            reachable: true,
+            $or:[]
         };
-        query['dependencies.' + data.taskId] = {
+ 
+        var dependenciesItr = {};
+        dependenciesItr['dependencies.' + data.taskId] = {
             $in: _.difference(Constants.Task.FinishedStates, [data.state])
         };
+        query.$or.push(dependenciesItr);
+
+        delete dependenciesItr['dependencies.' + data.taskId];
+        dependenciesItr['dependencies.anyOf.' + data.taskId] = {
+            $in: _.difference(Constants.Task.FinishedStates, [data.state])
+        };
```

Comment:
If only one of dependency criteria within `anyOf` is unreachable, this doesn't mean the whole task is unreachable, for example:
```javascript
waitOn: {
  anyOf: {
    'foo': 'succeed',
    'bar': 'succeed'
  }
}
```

If the task `foo` is failed, the criteria `'foo': 'succeed'` is surely never been meet, but the task `bar` may succeed, so the criteria `'bar': 'succeed'` is still valid here, thus the whole task is still reachable.